File: src/TabWriter.zig
const std = @import("std");

pub const Alignment = enum { left, right };

pub const Column = struct {
    width: usize,
    alignment: Alignment = .left,
};

pub const Error = error{InvalidColumnCount};

pub const TabWriter = struct {
    writer: *std.Io.Writer,
    columns: []const Column,
    gap: []const u8 = "  ",

    pub fn init(writer: *std.Io.Writer, columns: []const Column) TabWriter {
        return .{ .writer = writer, .columns = columns };
    }

    pub fn withGap(writer: *std.Io.Writer, columns: []const Column, gap: []const u8) TabWriter {
        return .{ .writer = writer, .columns = columns, .gap = gap };
    }

    pub fn writeHeader(self: *TabWriter, headers: []const []const u8) !void {
        try self.writeRow(headers);
    }

    pub fn writeSeparator(self: *TabWriter, ch: u8) !void {
        for (self.columns, 0..) |column, idx| {
            if (idx != 0) try self.writer.writeAll(self.gap);
            const width = column.width;
            const segment_width = if (width == 0) 3 else width;
            try writeRepeated(self.writer, segment_width, ch);
        }
        try self.writer.writeByte('\n');
    }

    pub fn writeRow(self: *TabWriter, values: []const []const u8) !void {
        if (values.len != self.columns.len) {
            return Error.InvalidColumnCount;
        }

        for (values, self.columns, 0..) |value, column, idx| {
            if (idx != 0) try self.writer.writeAll(self.gap);
            try writeColumn(self.writer, column, value);
        }
        try self.writer.writeByte('\n');
    }

    fn writeColumn(writer: *std.Io.Writer, column: Column, value: []const u8) !void {
        const width = column.width;
        if (width == 0 or value.len >= width) {
            try writer.writeAll(value);
            return;
        }

        const padding = width - value.len;
        switch (column.alignment) {
            .left => {
                try writer.writeAll(value);
                try writeRepeated(writer, padding, ' ');
            },
            .right => {
                try writeRepeated(writer, padding, ' ');
                try writer.writeAll(value);
            },
        }
    }
};

fn writeRepeated(writer: *std.Io.Writer, count: usize, ch: u8) !void {
    if (count == 0) return;

    var buffer: [16]u8 = undefined;
    @memset(&buffer, ch);

    var remaining = count;
    while (remaining >= buffer.len) : (remaining -= buffer.len) {
        try writer.writeAll(buffer[0..]);
    }

    if (remaining > 0) {
        try writer.writeAll(buffer[0..remaining]);
    }
}

File: cpp/src/tab_writer.cpp
#include "tab_writer.hpp"
#include <iomanip>
#include <sstream>

namespace wtfs {

TabWriter::TabWriter(std::ostream& out, const std::vector<Column>& columns, TableStyle style)
    : out_(out), columns_(columns), style_(style) {
}

void TabWriter::write_header() {
    if (header_written_) return;

    if (style_ == TableStyle::Box) {
        out_ << "┌";
        for (size_t i = 0; i < columns_.size(); ++i) {
            if (i > 0) out_ << "─┬";
            out_ << std::string(columns_[i].width + 2, '─');
        }
        out_ << "─┐\n";
    }

    // Write header row
    if (style_ == TableStyle::Box) {
        out_ << "│";
    }

    for (size_t i = 0; i < columns_.size(); ++i) {
        if (i > 0 && style_ == TableStyle::Box) {
            out_ << " │";
        }
        out_ << " " << pad_cell(columns_[i].header, columns_[i].width, Alignment::Left);
    }

    if (style_ == TableStyle::Box) {
        out_ << " │";
    }
    out_ << "\n";

    // Write separator
    if (style_ == TableStyle::Box) {
        out_ << "├";
        for (size_t i = 0; i < columns_.size(); ++i) {
            if (i > 0) out_ << "─┼";
            out_ << std::string(columns_[i].width + 2, '─');
        }
        out_ << "─┤\n";
    } else {
        write_separator('-');
    }

    header_written_ = true;
}

void TabWriter::write_separator(char fill) {
    if (style_ == TableStyle::Box) {
        out_ << "├";
        for (size_t i = 0; i < columns_.size(); ++i) {
            if (i > 0) out_ << fill << "┼";
            out_ << std::string(columns_[i].width + 2, fill);
        }
        out_ << fill << "┤\n";
    } else {
        for (size_t i = 0; i < columns_.size(); ++i) {
            if (i > 0) out_ << " ";
            out_ << std::string(columns_[i].width + 2, fill);
        }
        out_ << "\n";
    }
}

void TabWriter::write_row(const std::vector<std::string>& cells) {
    if (!header_written_) {
        write_header();
    }

    if (style_ == TableStyle::Box) {
        out_ << "│";
    }

    for (size_t i = 0; i < columns_.size() && i < cells.size(); ++i) {
        if (i > 0 && style_ == TableStyle::Box) {
            out_ << " │";
        }
        out_ << " " << pad_cell(cells[i], columns_[i].width, columns_[i].alignment);
    }

    if (style_ == TableStyle::Box) {
        out_ << " │";
    }
    out_ << "\n";
}

void TabWriter::finish() {
    if (finished_) return;

    if (style_ == TableStyle::Box) {
        out_ << "└";
        for (size_t i = 0; i < columns_.size(); ++i) {
            if (i > 0) out_ << "─┴";
            out_ << std::string(columns_[i].width + 2, '─');
        }
        out_ << "─┘\n";
    }

    finished_ = true;
}

std::string TabWriter::pad_cell(const std::string& content, size_t width, Alignment align) const {
    if (content.length() >= width) {
        return content.substr(0, width);
    }

    size_t padding = width - content.length();

    switch (align) {
        case Alignment::Left:
            return content + std::string(padding, ' ');

        case Alignment::Right:
            return std::string(padding, ' ') + content;

        case Alignment::Center: {
            size_t left_pad = padding / 2;
            size_t right_pad = padding - left_pad;
            return std::string(left_pad, ' ') + content + std::string(right_pad, ' ');
        }
    }

    return content;
}

} // namespace wtfs
